File: examples_java/src/com/libzupt/ExampleFile.java
package com.libzupt;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * File Encryption/Decryption Example.
 * Shows how to encrypt and decrypt files.
 */
public class ExampleFile {

    public static void main(String[] args) throws IOException {
        System.out.println("=".repeat(60));
        System.out.println("libzupt - File Encryption/Decryption Example");
        System.out.println("=".repeat(60));
        System.out.println();

        // Generate key pair
        System.out.println("1. Generating key pair...");
        KeyGenerator keygen = new KeyGenerator();
        KeyPair keypair = keygen.generateKeyPair();
        System.out.println("   Key pair generated");
        System.out.println();

        // Create temporary test file
        Path tempDir = Files.createTempDirectory("zupt_test");
        Path testFile = tempDir.resolve("test.txt");
        Path cipherFile = tempDir.resolve("test.txt.enc");

        try {
            String originalContent =
                "This is a secret text file.\n" +
                "Line 2: Contains sensitive information.\n" +
                "Line 3: End of file.\n";
            Files.writeString(testFile, originalContent);

            System.out.println("2. Created test file: " + testFile);
            System.out.println("   Original content:");
            System.out.println(originalContent);

            // Encrypt file
            Encryptor encryptor = new Encryptor(keypair.publicKey);
            Object[] encrypted = encryptor.encryptFile(testFile.toString());
            byte[] ciphertext = (byte[]) encrypted[0];
            byte[] encHeader = (byte[]) encrypted[1];

            System.out.println("3. Encrypting file...");
            System.out.println("   Ciphertext size: " + ciphertext.length + " bytes");
            System.out.println("   Header size: " + encHeader.length + " bytes");
            System.out.println();

            // Save ciphertext
            Files.write(cipherFile, ciphertext);
            System.out.println("4. Saved ciphertext to: " + cipherFile);
            System.out.println();

            // Decrypt file
            Decryptor decryptor = new Decryptor(keypair.secretKey);
            byte[] decrypted = decryptor.decryptFile(cipherFile.toString(), encHeader);
            String decryptedContent = new String(decrypted);

            System.out.println("5. Decrypting file...");
            System.out.println("   Decrypted size: " + decrypted.length + " bytes");
            System.out.println("   Decrypted content:");
            System.out.println(decryptedContent);

            // Verify
            if (!originalContent.equals(decryptedContent)) {
                throw new RuntimeException("Decryption failed!");
            }
            System.out.println();

            System.out.println("6. Cleaned up temporary files");

        } finally {
            // Cleanup
            try {
                Files.deleteIfExists(cipherFile);
                Files.deleteIfExists(testFile);
                Files.deleteIfExists(tempDir);
            } catch (IOException e) {
                // Ignore cleanup errors
            }
        }

        System.out.println("=".repeat(60));
        System.out.println("File encryption/decryption example passed!");
        System.out.println("=".repeat(60));
    }
}

File: examples_rust/examples/example_file.rs
use libzupt_rust;
use std::fs;
use std::io::Write;

fn main() {
    println!("{}", "=".repeat(60));
    println!("libzupt - File Encryption/Decryption Example (Rust)");
    println!("{}", "=".repeat(60));
    println!();

    println!("1. Generating key pair...");
    let keypair = libzupt_rust::generate_keypair().expect("Failed to generate key pair");
    println!("   Key pair generated");
    println!();

    let test_content = b"This is a secret text file.\nLine 2: Contains sensitive information.\nLine 3: End of file.\n";
    let test_path = "rust_test_file.txt";
    let cipher_path = "rust_test_file.txt.enc";

    {
        let mut f = fs::File::create(test_path).expect("Failed to create test file");
        f.write_all(test_content).expect("Failed to write test file");
    }
    println!("2. Created test file: {}", test_path);
    println!("   Original content:\n{}", String::from_utf8_lossy(test_content));

    println!("3. Encrypting file...");
    let (ciphertext, enc_header) =
        libzupt_rust::encrypt_file(&keypair.public_key, test_path).expect("Failed to encrypt file");
    println!("   Ciphertext size: {} bytes", ciphertext.len());
    println!("   Header size: {} bytes", enc_header.len());
    println!();

    libzupt_rust::write_file(cipher_path, &ciphertext).expect("Failed to write cipher file");
    println!("4. Saved ciphertext to: {}", cipher_path);
    println!();

    println!("5. Decrypting file...");
    let decrypted = libzupt_rust::decrypt_file(&keypair.secret_key, cipher_path, &enc_header)
        .expect("Failed to decrypt file");
    println!("   Decrypted size: {} bytes", decrypted.len());
    println!(
        "   Decrypted content:\n{}",
        String::from_utf8_lossy(&decrypted)
    );

    assert_eq!(decrypted, test_content, "Decryption failed!");
    println!();

    fs::remove_file(test_path).ok();
    fs::remove_file(cipher_path).ok();
    println!("6. Cleaned up temporary files");
    println!();

    println!("{}", "=".repeat(60));
    println!("File encryption/decryption example passed!");
    println!("{}", "=".repeat(60));
}
